File: sintactico/Lexico.py
class Lexico:
    def __init__(self, e=None):
        self.indice = 0
        self.simbolo_per = True
        self.fuente = e if e else ""
        self.estado = 0
        self.continuar = True
        self.simbolo = ""
        self.tipo = 0
        self.c = ''

    def entrada(self, e):
        self.indice = 0
        self.fuente = e

    def tipo_acad(self, tipo):
        simbolo_tipo = ""
        if tipo == 0:
            simbolo_tipo = "ID"
        elif tipo == 1:
            simbolo_tipo = "Entero"
        elif tipo == 2:
            simbolo_tipo = "Real"
        elif tipo == 3:
            simbolo_tipo = "Cadena"
        elif tipo == 4:
            simbolo_tipo = "TipoDato"
        elif tipo == 5:
            simbolo_tipo = "OpAdicion"
        elif tipo == 6:
            simbolo_tipo = "OpMul"
        elif tipo == 7:
            simbolo_tipo = "OpRelacional"
        elif tipo == 8:
            simbolo_tipo = "OpOr"
        elif tipo == 9:
            simbolo_tipo = "OpAnd"
        elif tipo == 10:
            simbolo_tipo = "OpNot"
        elif tipo == 11:
            simbolo_tipo = "OpComparacion"
        elif tipo == 12:
            simbolo_tipo = "PuntoComa"
        elif tipo == 13:
            simbolo_tipo = "Coma"
        elif tipo == 14:
            simbolo_tipo = "ParentesisIzq"
        elif tipo == 15:
            simbolo_tipo = "ParentesisDer"
        elif tipo == 16:
            simbolo_tipo = "LlaveIzq"
        elif tipo == 17:
            simbolo_tipo = "LlaveDer"
        elif tipo == 18:
            simbolo_tipo = "OpAsignacion"
        elif tipo == 19:
            simbolo_tipo = "IF"
        elif tipo == 20:
            simbolo_tipo = "WHILE"
        elif tipo == 21:
            simbolo_tipo = "RETURN"
        elif tipo == 22:
            simbolo_tipo = "ELSE"
        elif tipo == 23:
            simbolo_tipo = "FIN"
        else:
            simbolo_tipo = "ERROR"
        return simbolo_tipo

    def sig_simbolo(self):
        self.estado = 0
        self.continuar = True
        self.simbolo = ""

        while self.continuar:
            if self.simbolo_per:
                self.c = self.sig_caracter()
            self.simbolo_per = True

            if self.estado == 0:
                if self.es_letra(self.c):
                    self.sig_estado(1)
                elif self.es_digito(self.c):
                    self.sig_estado(3)
                elif self.es_adicion(self.c):
                    self.aceptacion(5)
                elif self.es_mul(self.c):
                    self.aceptacion(6)
                elif self.es_mayor_que(self.c) or self.es_menor_que(self.c):
                    self.sig_estado(6)
                elif self.es_or(self.c):
                    self.sig_estado(7)
                elif self.es_and(self.c):
                    self.sig_estado(8)
                elif self.es_exclamacion(self.c):
                    self.sig_estado(9)
                elif self.es_igual_que(self.c):
                    self.sig_estado(10)
                elif self.es_punto_coma(self.c):
                    self.aceptacion(12)
                elif self.es_coma(self.c):
                    self.aceptacion(13)
                elif self.es_parentesis_izq(self.c):
                    self.aceptacion(14)
                elif self.es_parentesis_der(self.c):
                    self.aceptacion(15)
                elif self.es_llave_izq(self.c):
                    self.aceptacion(16)
                elif self.es_llave_der(self.c):
                    self.aceptacion(17)
                elif self.es_pesos(self.c):
                    self.aceptacion(23)
                elif self.es_espacio(self.c):
                    self.sig_estado(0)
                elif self.es_comillas(self.c):
                    self.sig_estado(11)
                else:
                    self.aceptacion(-1)

            elif self.estado == 1:
                if self.es_letra(self.c):
                    self.sig_estado(1)
                elif self.es_digito(self.c):
                    self.sig_estado(2)
                else:
                    self.simbolo_per = False
                    self.aceptacion(self.rev_cadena())

            elif self.estado == 2:
                if self.es_letra(self.c) or self.es_digito(self.c):
                    self.sig_estado(2)
                else:
                    self.simbolo_per = False
                    self.aceptacion(0)

            elif self.estado == 3:
                if self.es_digito(self.c):
                    self.sig_estado(3)
                elif self.es_punto(self.c):
                    self.sig_estado(4)
                else:
                    self.simbolo_per = False
                    self.aceptacion(1)

            elif self.estado == 4:
                if self.es_digito(self.c):
                    self.sig_estado(5)
                else:
                    self.simbolo_per = False
                    self.aceptacion(-1)

            elif self.estado == 5:
                if self.es_digito(self.c):
                    self.sig_estado(5)
                else:
                    self.simbolo_per = False
                    self.aceptacion(2)

            elif self.estado == 6:
                if self.es_igual_que(self.c):
                    self.aceptacion(7)
                else:
                    self.simbolo_per = False
                    self.aceptacion(7)

            elif self.estado == 7:
                if self.es_or(self.c):
                    self.aceptacion(8)
                else:
                    self.simbolo_per = False
                    self.aceptacion(-1)

            elif self.estado == 8:
                if self.es_and(self.c):
                    self.aceptacion(9)
                else:
                    self.simbolo_per = False
                    self.aceptacion(-1)

            elif self.estado == 9:
                if self.es_igual_que(self.c):
                    self.aceptacion(11)
                else:
                    self.simbolo_per = False
                    self.aceptacion(10)

            elif self.estado == 10:
                if self.es_igual_que(self.c):
                    self.aceptacion(11)
                else:
                    self.simbolo_per = False
                    self.aceptacion(18)

            elif self.estado == 11:
                if self.es_comillas(self.c):
                    self.aceptacion(3)
                elif self.isascii(self.c):
                    self.sig_estado(11)

        tipo = 0
        if self.estado == 0:
            tipo = 0
        elif self.estado == 1:
            tipo = 1
        elif self.estado == 2:
            tipo = 2
        elif self.estado == 3:
            tipo = 3
        elif self.estado == 4:
            tipo = 4
        elif self.estado == 5:
            tipo = 5
        elif self.estado == 6:
            tipo = 6
        elif self.estado == 7:
            tipo = 7
        elif self.estado == 8:
            tipo = 8
        elif self.estado == 9:
            tipo = 9
        elif self.estado == 10:
            tipo = 10
        elif self.estado == 11:
            tipo = 11
        elif self.estado == 12:
            tipo = 12
        elif self.estado == 13:
            tipo = 13
        elif self.estado == 14:
            tipo = 14
        elif self.estado == 15:
            tipo = 15
        elif self.estado == 16:
            tipo = 16
        elif self.estado == 17:
            tipo = 17
        elif self.estado == 18:
            tipo = 18
        elif self.estado == 19:
            tipo = 19
        elif self.estado == 20:
            tipo = 20
        elif self.estado == 21:
            tipo = 21
        elif self.estado == 22:
            tipo = 22
        elif self.estado == 23:
            tipo = 23
        else:
            tipo = -1

        return tipo

    def terminado(self):
        return self.indice >= len(self.fuente)

    def sig_caracter(self):
        c = self.fuente[self.indice]
        self.indice += 1
        if self.terminado():
            return '$'
        return c

    def sig_estado(self, e):
        self.estado = e

        if self.simbolo_per:
            self.simbolo += self.c

    def aceptacion(self, e):
        self.sig_estado(e)
        self.continuar = False

    def retroceso(self):
        if self.c != '$':
            self.indice -= 1
        self.continuar = False

    def rev_cadena(self):
        if self.simbolo == "int" or self.simbolo == "float" or self.simbolo == "void":
            return 4
        elif self.simbolo == "if":
            return 19
        elif self.simbolo == "while":
            return 20
        elif self.simbolo == "return":
            return 21
        elif self.simbolo == "else":
            return 22
        else:
            return 0

    def es_letra(self, c):
        return c.isalpha() or c == '_'

    def es_digito(self, c):
        return c.isdigit()

    def es_espacio(self, c):
        return c == ' ' or c == '\t' or c == '\n'

    def es_dos_puntos(self, c):
        return c == ':'

    def es_punto_coma(self, c):
        return c == ';'

    def es_coma(self, c):
        return c == ','

    def es_adicion(self, c):
        return c == '+' or c == '-'

    def es_mul(self, c):
        return c == '*' or c == '/'

    def es_parentesis_izq(self, c):
        return c == '('

    def es_parentesis_der(self, c):
        return c == ')'

    def es_llave_izq(self, c):
        return c == '{'

    def es_llave_der(self, c):
        return c == '}'

    def es_punto(self, c):
        return c == '.'

    def es_mayor_que(self, c):
        return c == '>'

    def es_menor_que(self, c):
        return c == '<'

    def es_igual_que(self, c):
        return c == '='

    def es_exclamacion(self, c):
        return c == '!'

    def es_or(self, c):
        return c == '|'

    def es_and(self, c):
        return c == '&'

    def es_pesos(self, c):
        return c == '$'

    def es_comillas(self, c):
        return c == '"'





class Lexico:
    ERROR = -1
    IDENTIFICADOR = 0
    ENTERO = 1
    REAL = 2
    CADENA = 3
    TIPO_FUNCION = 4
    OP_ADICION = 5
    OP_MUL = 6
    OP_RELACIONAL = 7
    OP_OR = 8
    OP_AND = 9
    OP_NOT = 10
    OP_IGUAL = 11
    PUNTOCOMA = 12
    COMA = 13
    PARENTESIS_IZQ = 14
    PARENTESIS_DER = 15
    LLAVE_IZQ = 16
    LLAVE_DER = 17
    OP_ASIGNACION = 18
    RES_IF = 19
    RES_WHILE = 20
    RES_RETURN = 21
    RES_ELSE = 22
    FIN = 23

    def __init__(self, e=None):
        self.simbolo = ""
        self.tipo = 0
        self.fuente = e if e else ""
        self.indice = 0
        self.continuar = True
        self.simbolo_per = True
        self.c = ""
        self.estado = 0

    def entrada(self, e):
        self.indice = 0
        self.fuente = e

    def tipo_acad(self, tipo):
        if tipo == 0:
            return "ID"
        elif tipo == 1:
            return "Entero"
        elif tipo == 2:
            return "Real"
        elif tipo == 3:
            return "Cadena"
        elif tipo == 4:
            return "TipoDato"
        elif tipo == 5:
            return "OpAdicion"
        elif tipo == 6:
            return "OpMul"
        elif tipo == 7:
            return "OpRelacional"
        elif tipo == 8:
            return "OpOr"
        elif tipo == 9:
            return "OpAnd"
        elif tipo == 10:
            return "OpNot"
        elif tipo == 11:
            return "OpComparacion"
        elif tipo == 12:
            return "PuntoComa"
        elif tipo == 13:
            return "Coma"
        elif tipo == 14:
            return "ParentesisIzq"
        elif tipo == 15:
            return "ParentesisDer"
        elif tipo == 16:
            return "LlaveIzq"
        elif tipo == 17:
            return "LlaveDer"
        elif tipo == 18:
            return "OpAsignacion"
        elif tipo == 19:
            return "IF"
        elif tipo == 20:
            return "WHILE"
        elif tipo == 21:
            return "RETURN"
        elif tipo == 22:
            return "ELSE"
        elif tipo == 23:
            return "FIN"
        else:
            return "ERROR"

    def sig_simbolo(self):
        self.estado = 0
        self.continuar = True
        self.simbolo = ""

        while self.continuar:
            if self.simbolo_per:
                self.c = self.sig_caracter()
            self.simbolo_per = True

            if self.estado == 0:
                if self.es_letra(self.c):
                    self.sig_estado(1)
                elif self.es_digito(self.c):
                    self.sig_estado(3)
                elif self.es_adicion(self.c):
                    self.aceptacion(5)
                elif self.es_mul(self.c):
                    self.aceptacion(6)
                elif self.es_mayor_que(self.c) or self.es_menor_que(self.c):
                    self.sig_estado(6)
                elif self.es_or(self.c):
                    self.sig_estado(7)
                elif self.es_and(self.c):
                    self.sig_estado(8)
                elif self.es_exclamacion(self.c):
                    self.sig_estado(9)
                elif self.es_igual_que(self.c):
                    self.sig_estado(10)
                elif self.es_punto_coma(self.c):
                    self.aceptacion(12)
                elif self.es_coma(self.c):
                    self.aceptacion(13)
                elif self.es_parentesis_izq(self.c):
                    self.aceptacion(14)
                elif self.es_parentesis_der(self.c):
                    self.aceptacion(15)
                elif self.es_llave_izq(self.c):
                    self.aceptacion(16)
                elif self.es_llave_der(self.c):
                    self.aceptacion(17)
                elif self.es_pesos(self.c):
                    self.aceptacion(23)
                elif self.es_espacio(self.c):
                    self.sig_estado(0)
                elif self.es_comillas(self.c):
                    self.sig_estado(11)
                else:
                    self.aceptacion(-1)

            elif self.estado == 1:
                if self.es_letra(self.c):
                    self.sig_estado(1)
                elif self.es_digito(self.c):
                    self.sig_estado(2)
                else:
                    self.simbolo_per = False
                    self.aceptacion(self.rev_cadena())

            elif self.estado == 2:
                if self.es_letra(self.c) or self.es_digito(self.c):
                    self.sig_estado(2)
                else:
                    self.simbolo_per = False
                    self.aceptacion(0)

            elif self.estado == 3:
                if self.es_digito(self.c):
                    self.sig_estado(3)
                elif self.es_punto(self.c):
                    self.sig_estado(4)
                else:
                    self.simbolo_per = False
                    self.aceptacion(1)

            elif self.estado == 4:
                if self.es_digito(self.c):
                    self.sig_estado(5)
                else:
                    self.simbolo_per = False
                    self.aceptacion(-1)

            elif self.estado == 5:
                if self.es_digito(self.c):
                    self.sig_estado(5)
                else:
                    self.simbolo_per = False
                    self.aceptacion(2)

            elif self.estado == 6:
                if self.es_igual_que(self.c):
                    self.aceptacion(7)
                else:
                    self.simbolo_per = False
                    self.aceptacion(7)

            elif self.estado == 7:
                if self.es_or(self.c):
                    self.aceptacion(8)
                else:
                    self.simbolo_per = False
                    self.aceptacion(-1)

            elif self.estado == 8:
                if self.es_and(self.c):
                    self.aceptacion(9)
                else:
                    self.simbolo_per = False
                    self.aceptacion(-1)

            elif self.estado == 9:
                if self.es_igual_que(self.c):
                    self.aceptacion(11)
                else:
                    self.simbolo_per = False
                    self.aceptacion(10)

            elif self.estado == 10:
                if self.es_igual_que(self.c):
                    self.aceptacion(11)
                else:
                    self.simbolo_per = False
                    self.aceptacion(18)

            elif self.estado == 11:
                if self.es_comillas(self.c):
                    self.aceptacion(3)
                elif self.isascii(self.c):
                    self.sig_estado(11)

        tipo = 0
        if self.estado == 0:
            tipo = self.IDENTIFICADOR
        elif self.estado == 1:
            tipo = self.ENTERO
        elif self.estado == 2:
            tipo = self.REAL
        elif self.estado == 3:
            tipo = self.CADENA
        elif self.estado == 4:
            tipo = self.TIPO_FUNCION
        elif self.estado == 5:
            tipo = self.OP_ADICION
        elif self.estado == 6:
            tipo = self.OP_MUL
        elif self.estado == 7:
            tipo = self.OP_RELACIONAL
        elif self.estado == 8:
            tipo = self.OP_OR
        elif self.estado == 9:
            tipo = self.OP_AND
        elif self.estado == 10:
            tipo = self.OP_NOT
        elif self.estado == 11:
            tipo = self.OP_IGUAL
        elif self.estado == 12:
            tipo = self.PUNTOCOMA
        elif self.estado == 13:
            tipo = self.COMA
        elif self.estado == 14:
            tipo = self.PARENTESIS_IZQ
        elif self.estado == 15:
            tipo = self.PARENTESIS_DER
        elif self.estado == 16:
            tipo = self.LLAVE_IZQ
        elif self.estado == 17:
            tipo = self.LLAVE_DER
        elif self.estado == 18:
            tipo = self.OP_ASIGNACION
        elif self.estado == 19:
            tipo = self.RES_IF
        elif self.estado == 20:
            tipo = self.RES_WHILE
        elif self.estado == 21:
            tipo = self.RES_RETURN
        elif self.estado == 22:
            tipo = self.RES_ELSE
        elif self.estado == 23:
            tipo = self.FIN
        else:
            tipo = self.ERROR

        return tipo

    def terminado(self):
        return self.indice >= len(self.fuente)

    def sig_caracter(self):
        c = self.fuente[self.indice]
        self.indice += 1
        if self.terminado():
            return '$'
        return c

    def sig_estado(self, e):
        self.estado = e

        if self.simbolo_per:
            self.simbolo += self.c

    def aceptacion(self, e):
        self.sig_estado(e)
        self.continuar = False

    def retroceso(self):
        if self.c != '$':
            self.indice -= 1
        self.continuar = False

    def rev_cadena(self):
        if self.simbolo == "int" or self.simbolo == "float" or self.simbolo == "void":
            return 4
        elif self.simbolo == "if":
            return 19
        elif self.simbolo == "while":
            return 20
        elif self.simbolo == "return":
            return 21
        elif self.simbolo == "else":
            return 22
        else:
            return 0

    def es_letra(self, c):
        return c.isalpha() or c == '_'

    def es_digito(self, c):
        return c.isdigit()

    def es_espacio(self, c):
        return c == ' ' or c == '\t' or c == '\n'

    def es_dos_puntos(self, c):
        return c == ':'

    def es_punto_coma(self, c):
        return c == ';'

    def es_coma(self, c):
        return c == ','

    def es_adicion(self, c):
        return c == '+' or c == '-'

    def es_mul(self, c):
        return c == '*' or c == '/'

    def es_parentesis_izq(self, c):
        return c == '('

    def es_parentesis_der(self, c):
        return c == ')'

    def es_llave_izq(self, c):
        return c == '{'

    def es_llave_der(self, c):
        return c == '}'

    def es_punto(self, c):
        return c == '.'

    def es_mayor_que(self, c):
        return c == '>'

    def es_menor_que(self, c):
        return c == '<'

    def es_igual_que(self, c):
        return c == '='

    def es_exclamacion(self, c):
        return c == '!'

    def es_or(self, c):
        return c == '|'

    def es_and(self, c):
        return c == '&'

    def es_pesos(self, c):
        return c == '$'

    def es_comillas(self, c):
        return c == '"'

File: sintactico/Lexico.java
package sintactico;

public class Lexico {
    // Constantes de tipo
    static final int ERROR = -1;
    static final int IDENTIFICADOR = 0;
    static final int ENTERO = 1;
    static final int REAL = 2;
    static final int CADENA = 3;
    static final int TIPO_FUNCION = 4;
    static final int OP_ADICION = 5;
    static final int OP_MUL = 6;
    static final int OP_RELACIONAL = 7;
    static final int OP_OR = 8;
    static final int OP_AND = 9;
    static final int OP_NOT = 10;
    static final int OP_IGUAL = 11;
    static final int PUNTOCOMA = 12;
    static final int COMA = 13;
    static final int PARENTESIS_IZQ = 14;
    static final int PARENTESIS_DER = 15;
    static final int LLAVE_IZQ = 16;
    static final int LLAVE_DER = 17;
    static final int OP_ASIGNACION = 18;
    static final int RES_IF = 19;
    static final int RES_WHILE = 20;
    static final int RES_RETURN = 21;
    static final int RES_ELSE = 22;
    static final int FIN = 23;

    String simbolo;
    int tipo;
    String fuente;

    int indice;
    boolean continuar;
    boolean simbolo_per;
    char c;
    int estado;

    // Constructor
    public Lexico() {
        indice = 0;
        simbolo_per = true;
    }

    // Constructor con parámetro para la entrada
    public Lexico(String e) {
        indice = 0;
        simbolo_per = true;
        fuente = e;
    }

    // Método para establecer la entrada
    public void setEntrada(String e) {
        indice = 0;
        fuente = e;
    }

    // Método para obtener el tipo académico de un tipo numérico
    public String getTipoAcademico(int tipo) {
        String simbolo_tipo;

        switch (tipo) {
            case IDENTIFICADOR:
                simbolo_tipo = "ID";
                break;
            case ENTERO:
                simbolo_tipo = "Entero";
                break;
            case REAL:
                simbolo_tipo = "Real";
                break;
            case CADENA:
                simbolo_tipo = "Cadena";
                break;
            case TIPO_FUNCION:
                simbolo_tipo = "TipoDato";
                break;
            case OP_ADICION:
                simbolo_tipo = "OpAdicion";
                break;
            case OP_MUL:
                simbolo_tipo = "OpMul";
                break;
            case OP_RELACIONAL:
                simbolo_tipo = "OpRelacional";
                break;
            case OP_OR:
                simbolo_tipo = "OpOr";
                break;
            case OP_AND:
                simbolo_tipo = "OpAnd";
                break;
            case OP_NOT:
                simbolo_tipo = "OpNot";
                break;
            case OP_IGUAL:
                simbolo_tipo = "OpComparacion";
                break;
            case PUNTOCOMA:
                simbolo_tipo = "PuntoComa";
                break;
            case COMA:
                simbolo_tipo = "Coma";
                break;
            case PARENTESIS_IZQ:
                simbolo_tipo = "ParentesisIzq";
                break;
            case PARENTESIS_DER:
                simbolo_tipo = "ParentesisDer";
                break;
            case LLAVE_IZQ:
                simbolo_tipo = "LlaveIzq";
                break;
            case LLAVE_DER:
                simbolo_tipo = "LlaveDer";
                break;
            case OP_ASIGNACION:
                simbolo_tipo = "OpAsignacion";
                break;
            case RES_IF:
                simbolo_tipo = "IF";
                break;
            case RES_WHILE:
                simbolo_tipo = "WHILE";
                break;
            case RES_RETURN:
                simbolo_tipo = "RETURN";
                break;
            case RES_ELSE:
                simbolo_tipo = "ELSE";
                break;
            case FIN:
                simbolo_tipo = "FIN";
                break;
            default:
                simbolo_tipo = "ERROR";
                break;
        }
        return simbolo_tipo;
    }

    // Método para obtener el siguiente símbolo del análisis léxico
    public int siguienteSimbolo() {
        estado = 0;
        continuar = true;
        simbolo = "";

        while (continuar) {
            if (simbolo_per)
                c = siguienteCaracter();
            simbolo_per = true;

            switch (estado) {
                case 0:
                    if (esLetra(c))
                        cambiarEstado(1);
                    else if (esDigito(c))
                        cambiarEstado(3);
                    else if (esAdicion(c))
                        aceptar(5);
                    else if (esMul(c))
                        aceptar(6);
                    else if (esMayorQue(c) || esMenorQue(c))
                        cambiarEstado(6);
                    else if (esOr(c))
                        cambiarEstado(7);
                    else if (esAnd(c))
                        cambiarEstado(8);
                    else if (esExclamacion(c))
                        cambiarEstado(9);
                    else if (esIgualQue(c))
                        cambiarEstado(10);
                    else if (esPuntoComa(c))
                        aceptar(12);
                    else if (esComa(c))
                        aceptar(13);
                    else if (esParentesisIzq(c))
                        aceptar(14);
                    else if (esParentesisDer(c))
                        aceptar(15);
                    else if (esLlaveIzq(c))
                        aceptar(16);
                    else if (esLlaveDer(c))
                        aceptar(17);
                    else if (esPesos(c))
                        aceptar(23);
                    else if (esEspacio(c))
                        cambiarEstado(0);
                    else if (esComillas(c))
                        cambiarEstado(11);
                    else
                        aceptar(-1);
                    break;

                case 1:
                    if (esLetra(c))
                        cambiarEstado(1);
                    else if (esDigito(c))
                        cambiarEstado(2);
                    else {
                        simbolo_per = false;
                        aceptar(revisarCadena());
                    }
                    break;

                case 2:
                    if (esLetra(c) || esDigito(c))
                        cambiarEstado(2);
                    else {
                        simbolo_per = false;
                        aceptar(0);
                    }
                    break;

                case 3:
                    if (esDigito(c))
                        cambiarEstado(3);
                    else if (esPunto(c))
                        cambiarEstado(4);
                    else {
                        simbolo_per = false;
                        aceptar(1);
                    }
                    break;

                case 4:
                    if (esDigito(c))
                        cambiarEstado(5);
                    else {
                        simbolo_per = false;
                        aceptar(-1);
                    }
                    break;

                case 5:
                    if (esDigito(c))
                        cambiarEstado(5);
                    else {
                        simbolo_per = false;
                        aceptar(2);
                    }
                    break;

                case 6:
                    if (esIgualQue(c))
                        aceptar(7);
                    else {
                        simbolo_per = false;
                        aceptar(7);
                    }
                    break;

                case 7:
                    if (esOr(c))
                        aceptar(8);
                    else {
                        simbolo_per = false;
                        aceptar(-1);
                    }
                    break;

                case 8:
                    if (esAnd(c))
                        aceptar(9);
                    else {
                        simbolo_per = false;
                        aceptar(-1);
                    }
                    break;

                case 9:
                    if (esIgualQue(c))
                        aceptar(11);
                    else {
                        simbolo_per = false;
                        aceptar(10);
                    }
                    break;

                case 10:
                    if (esIgualQue(c))
                        aceptar(11);
                    else {
                        simbolo_per = false;
                        aceptar(18);
                    }
                    break;

                case 11:
                    if (esComillas(c))
                        aceptar(3);
                    else if (c >= 0 && c <= 127) {
                        cambiarEstado(11);
                    }
                    break;

                default:
                    break;
            }
        }

        switch (estado) {
            case 0:
                tipo = IDENTIFICADOR;
                break;

            case 1:
                tipo = ENTERO;
                break;

            case 2:
                tipo = REAL;
                break;

            case 3:
                tipo = CADENA;
                break;

            case 4:
                tipo = TIPO_FUNCION;
                break;

            case 5:
                tipo = OP_ADICION;
                break;

            case 6:
                tipo = OP_MUL;
                break;

            case 7:
                tipo = OP_RELACIONAL;
                break;

            case 8:
                tipo = OP_OR;
                break;

            case 9:
                tipo = OP_AND;
                break;

            case 10:
                tipo = OP_NOT;
                break;

            case 11:
                tipo = OP_IGUAL;
                break;

            case 12:
                tipo = PUNTOCOMA;
                break;

            case 13:
                tipo = COMA;
                break;

            case 14:
                tipo = PARENTESIS_IZQ;
                break;

            case 15:
                tipo = PARENTESIS_DER;
                break;

            case 16:
                tipo = LLAVE_IZQ;
                break;

            case 17:
                tipo = LLAVE_DER;
                break;

            case 18:
                tipo = OP_ASIGNACION;
                break;

            case 19:
                tipo = RES_IF;
                break;

            case 20:
                tipo = RES_WHILE;
                break;

            case 21:
                tipo = RES_RETURN;
                break;

            case 22:
                tipo = RES_ELSE;
                break;

            case 23:
                tipo = FIN;
                break;

            default:
                tipo = ERROR;
        }

        return tipo;
    }

    // Método para verificar si la entrada ha terminado
    public boolean terminado() {
        return indice >= fuente.length();
    }

    // Método para obtener el siguiente caracter de la entrada
    public char siguienteCaracter() {
        char c = fuente.charAt(indice++);
        if (terminado())
            return '$';
        return c;
    }

    // Método para cambiar al siguiente estado del autómata
    public void cambiarEstado(int e) {
        estado = e;

        if (simbolo_per)
            simbolo += c;
    }

    // Método para procesar la aceptación de un estado
    public void aceptar(int e) {
        cambiarEstado(e);
        continuar = false;
    }

    // Método para retroceder en la entrada
    public void retroceder() {
        if (c != '$')
            indice--;
        continuar = false;
    }

    // Método para obtener el tipo académico de una cadena
    public int revisarCadena() {
        if (simbolo.equals("int") || simbolo.equals("float") || simbolo.equals("void"))
            return 4;
        else if (simbolo.equals("if"))
            return 19;
        else if (simbolo.equals("while"))
            return 20;
        else if (simbolo.equals("return"))
            return 21;
        else if (simbolo.equals("else"))
            return 22;
        else
            return 0;
    }

    // Métodos para verificar tipos de caracteres
    public boolean esLetra(char c) {
        return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
    }

    public boolean esDigito(char c) {
        return (c >= '1' && c <= '9') || c == '0';
    }

    public boolean esEspacio(char c) {
        return c == ' ' || c == '\t' || c == '\n';
    }

    public boolean esDosPuntos(char c) {
        return c == ':';
    }

    public boolean esPuntoComa(char c) {
        return c == ';';
    }

    public boolean esComa(char c) {
        return c == ',';
    }

    public boolean esAdicion(char c) {
        return c == '+' || c == '-';
    }

    public boolean esMul(char c) {
        return c == '*' || c == '/';
    }

    public boolean esParentesisIzq(char c) {
        return c == '(';
    }

    public boolean esParentesisDer(char c) {
        return c == ')';
    }

    public boolean esLlaveIzq(char c) {
        return c == '{';
    }

    public boolean esLlaveDer(char c) {
        return c == '}';
    }

    public boolean esPunto(char c) {
        return c == '.';
    }

    public boolean esMayorQue(char c) {
        return c == '>';
    }

    public boolean esMenorQue(char c) {
        return c == '<';
    }

    public boolean esIgualQue(char c) {
        return c == '=';
    }

    public boolean esExclamacion(char c) {
        return c == '!';
    }

    public boolean esOr(char c) {
        return c == '|';
    }

    public boolean esAnd(char c) {
        return c == '&';
    }

    public boolean esPesos(char c) {
        return c == '$';
    }

    public boolean esComillas(char c) {
        return c == '"';
    }
}
